src/sys/windows/selector.rs

```diff
@@ -371,14 +411,233 @@ impl Selector {
+        /// `Registry::register`.
+        pub fn new(registry: &Registry, token: Token) -> Self {
+            let selector = poll::selector(registry);
+            Binding { selector: Arc::clone(&selector.inner), token }
+        }
+
+        /// Register a raw windows handle with the IOCP event loop.
+        ///
+        /// The underlying resource needs to support overlapped operations. Thus, files and similar
+        /// entities need to be opened in overlapped mode.
+        ///
+        /// While it is safely possible to call this function more than once it is discouraged to do
+        /// so as the handler will use the same token again when generating events. This makes it hard
+        /// for the user of the registered `Source`.
+        pub fn register_handle<H>(&self, handle: &H) -> io::Result<()> where H: AsRawHandle + ?Sized {
+            let handler = RawHandleCompletionHandler {
+                token: self.token
+            };
+            self.selector.cp.register_handle(handle, handler.into())
+        }
+
+        /// Inject an event to the event loop.
+        ///
+        /// In case it is required to generate an event even though there was no callback due to a
+        /// completed IO operation, this method can be used to mark the associated token as ready, e.g.
+        /// in case it is necessary to signal an initial readiness before any IO operation has been
+        /// scheduled.
+        pub fn inject_event(&self, readiness: Readiness) {
+            let event = readiness.into_event(self.token);
+            self.selector.injected_events.lock().unwrap().push(event)
+        }
+
+        /// Check whether the provided registry is the same as the one used during creation of the
+        /// binding.
+        ///
+        /// IOCP driven sources cannot unregister from their completion port before they get closed.
+        /// Therefore, implementations of reregister and deregister should check whether they have been
+        /// called on the same `Registry` using this method to report errors in case this constraint
+        /// has been violated.
+        pub fn is_same_registry(&self, registry: &Registry) -> bool {
+            Arc::ptr_eq(&self.selector, &poll::selector(registry).inner)
+        }
+    }
+
+    /// A wrapper around an internal instance over `miow::Overlapped` which is in
+    /// turn a wrapper around the Windows type `OVERLAPPED`.
+    ///
+    /// This type is required to be used for all IOCP operations on handles that are
+    /// registered with an event loop. The event loop will receive notifications
+    /// over `OVERLAPPED` pointers that have completed, and it will cast that
+    /// pointer to a pointer to this structure and invoke the associated callback.
+    #[repr(C)]
+    pub struct Overlapped {
+        inner: UnsafeCell<miow::Overlapped>,
+        #[cfg(debug_assertions)]
+        canary: u32,
+        callback: Box<dyn CompletionCallback>,
```

Comment:
Maybe `fn(&OVERLAPPED_ENTRY) -> Option<Readiness>` would be better, than we don't have to allocate.

Reply:
This is correct. I've chosen to use a boxed trait here because I was looking at mio_named_pipes which does some pointer magic to get its state back inside the callback functions fore read, write and connect. By using dynamic dispatch, we can avoid that and have more safe code, but if resource usage beats that argument (plus the "future hooks" argument from above) I'd be willing to change this to a pure function pointer.